modules/core/include/opencv2/core/hal/cvtype.h

```diff
@@ -0,0 +1,258 @@
+                                  ((cn) <= __CV_CN_LIMIT_EXP(3) ? 3 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(4) ? 4 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(5) ? 5 :          \
+                                  ((cn) <= __CV_CN_LIMIT_EXP(6) ? 6 : 7)))))))
+#define __CV_CN_EXPONENT(cn_bin)  (((cn_bin) >> CV_CN_BASE_LEN) & CV_SANITY_CN_EXP_MASK)
+#define __CV_CN_MAKE_BASE(cn)     (((cn) - 1) >> __CV_CN_MAKE_EXPONENT(cn)) /*& CV_SANITY_CN_BASE_MASK*/
+#define __CV_CN_BASE(cn_bin)      ((cn_bin) & CV_SANITY_CN_BASE_MASK)
+#define __CV_CN_DEFLATE(cn)       ((__CV_CN_MAKE_EXPONENT(cn) << CV_CN_BASE_LEN) | (__CV_CN_MAKE_BASE(cn) & CV_SANITY_CN_BASE_MASK)) /*((cn) - 1)*/
+#define __CV_CN_INFLATE__(e,b)    ((b + 1) << e)
+#define __CV_CN_INFLATE(cn_bin)   __CV_CN_INFLATE__(__CV_CN_EXPONENT(cn_bin), __CV_CN_BASE(cn_bin)) /*((cn_bin) + 1)*/
+
+//ElemType register pack/unpack operations (private)
+#define __CV_DEPTH_PACK(depth)    (((depth) & CV_SANITY_DEPTH_MASK) << CV_DEPTH_SHIFT)
+#define __CV_DEPTH_UNPACK(flags)  (((flags) & CV_MAT_DEPTH_MASK)  >> CV_DEPTH_SHIFT)
+#define __CV_CN_PACK(cn)          ((__CV_CN_DEFLATE(cn) & CV_SANITY_CN_MASK) << CV_CN_SHIFT)
+#define __CV_CN_UNPACK(flags)     __CV_CN_INFLATE(((flags) & CV_MAT_CN_MASK) >> CV_CN_SHIFT)
+/*
+TODO: static assertion on bad input channel
+TODO: static assertion on cn == cn to avoid random multiple evaluation*/
+#define __CV_TYPE_PACK(depth,cn)  (__CV_DEPTH_PACK(depth) | __CV_CN_PACK(cn))
+/*TODO: Rewite __CV_TYPE_UNPACK*/
+#define __CV_TYPE_UNPACK(flags)   __CV_TYPE_PACK(__CV_DEPTH_UNPACK(flags),__CV_CN_UNPACK(flags)) /*((flags) & CV_MAT_TYPE_MASK)*/
+
+//ElemType register - user operations (public)
+#define CV_CN_FIT(cn)             __CV_CN_INFLATE__(__CV_CN_MAKE_EXPONENT(cn), __CV_CN_MAKE_BASE(cn))
+#define CV_MAT_TYPE(flags)        __CV_TYPE_UNPACK(flags)
+#define CV_MAT_CN(flags)          __CV_CN_UNPACK(flags)
+#define CV_MAT_DEPTH(flags)       __CV_DEPTH_UNPACK(flags)
+#define CV_MAKETYPE(depth,cn)     __CV_TYPE_PACK(depth,cn)
+#define CV_MAKE_TYPE              CV_MAKETYPE
+
+//depth values (public)
+#define CV_8U        0
+#define CV_16U       2
+#define CV_32U       8
+#define CV_64U       9
+
+#define CV_8S        1
+#define CV_16S       3
+#define CV_32S       4
+#define CV_64S       11
+
+#define CV_16F       7
+#define CV_32F       5
+#define CV_64F       6
+
+#define CV_RAW       29
+#define CV_AUTO      30
+#define CV_UNDEF     31
+
+/*
+ * related #12732
+ *
+ * FIXME: Needs mapping from depth to actual type!
+ * This would help avoiding depth comparsions, and building arrays similar to the following
+ * Currently: only `DataType` exists, which maps from native type to depth
+ */
+
+//a register for sizeof depth values (private)
+static const int __cv_depth_sizeof[CV_DEPTH_MAX] = {
+    1/*CV_8U*/  , 1/*CV_8S*/ , 2/*CV_16U*/ , 2/*CV_16S*/  ,
```

Comment:
It is not good to declare/define tables in header files.

Reply:
Fixed! Anything else?